### ISSUE TITLE
Symbol's value as variable is void: zmq-REQ

### ISSUE THREAD

**Issue by prittjam** [dzop/emacs-jupyter#22]
When I try to install the package, I get this error and the package install fails. I am using the cask method of install. During installation I get several of these errors, and then when I try to use the package, it also reports the same error. On emacs startup it cannot find the zeromq module. It then asks to build it, and fails with this error.

**Comment by prittjam** [comment]
I am using the cask method of install. During installation I get several of these errors, and then when I try to use the package, it also reports the same error :(

**Comment by prittjam** [comment]
Autotools wasn't installed. Build of ZeroMQ failed as a result.

**Comment by Salompas** [comment]
I am having the same issue. I am trying to install on MacOS, via MELPA. After installing `emacs-jupyter` I am asked if I want to compile `zmq`, which I do. But at the end I get this error:

```
In toplevel form:
zmq.el:34:1:Error: Cannot open load file: cl-lib
make: *** [zmq.elc] Error 1

Compilation exited abnormally with code 2 at Wed Feb 13 18:03:51
```
And then trying to add `jupyter` to `org-babel-load-languages` leads to the same issue found by @prittjam 
Does anyone have advice? I'm pretty noob with these things. Thanks!

**Comment by dzop** [comment]
That's weird, does typing `M-: (require 'cl-lib) RET` give you an error? `cl-lib` is already built in to Emacs so I don't see how you would be getting that error.

Note you should be running Emacs 26 compiled with module support. If you installed Emacs using [homebrew](https://brew.sh/) this should already be the case.

@prittjam's issue has to do with having autotools installed which you can also get using homebrew using `brew install autoconf`. You may also need to have the package [exec-path-from-shell](https://github.com/purcell/exec-path-from-shell) installed so that Emacs knows where to find the autotools programs.

**Comment by Salompas** [comment]
@dzop 
`(require 'cl-lib)` does not give me an error.
I believe I installed emacs from emacsformacosx website, because I had other quirks with the other emacs versions ([see this stackexchange](https://emacs.stackexchange.com/questions/37808/tramp-cannot-save-file/44055#44055)). The version is Emacs 26.1.
I tried installing cl-lib from Melpa and requiring it, and then reinstalling emacs-jupyter, but the compilation step for zmq still fails.
I also use `exec-path-from-shell` .

**Comment by dzop** [comment]
Can you also show the value of `module-file-suffix`. If it is `nil` then your Emacs wasn't built with module support.

About `cl-lib`: OK, it might be that the Emacs being used to byte compile the `zmq` library is the wrong one. Can you show the value of running the shell command `which emacs`, if it shows `/usr/bin/emacs` then that is the case. Then, before you try to build `zmq` you can run
```(setenv "EMACS" "<path>")```. This will cause the Makefile to use the Emacs at `<path>` for byte compiling. In the Makefile, the `EMACS` variable is just set to `emacs` so the default Emacs provided by macOS is probably being.

**Comment by Salompas** [comment]
@dzop 
I installed emacs with `brew cask install emacs` and then installed `emacs-jupyter`. I wasn't even asked to compile `zmq` and it seems to be working!

I also noticed it wasn't emacsformacosx that I was using, but actually the version from MacPorts.
Unfortunately, the emacs from brew does not support borderless fullscreen and the fonts are not that good.

**Comment by Salompas** [comment]
> Can you also show the value of `module-file-suffix`. If it is `nil` then your Emacs wasn't built with module support.
> 
> About `cl-lib`: OK, it might be that the Emacs being used to byte compile the `zmq` library is the wrong one. Can you show the value of running the shell command `which emacs`, if it shows `/usr/bin/emacs` then that is the case. Then, before you try to build `zmq` you can run
> `(setenv "EMACS" "<path>")`. This will cause the Makefile to use the Emacs at `<path>` for byte compiling. In the Makefile, the `EMACS` variable is just set to `emacs` so the default Emacs provided by macOS is probably being.

I'll try that, give me a minute.

**Comment by Salompas** [comment]
On the Emacs version I was originally using (the one that couldn't compile zqm):
- The value of `module-file-suffix` is `.so`;
- Before installing the emacs with `brew` I believe `which emacs` was pointing to the Emacs that comes with MacOS, which was the 22.1.1 version;
- If someone else has the same problem (probably me in the future) should they set `(setenv "EMACS" "/Applications/Emacs.app/Contents/MacOS/Emacs")`?

The funny thing is, now that I have the brew emacs installed and working with jupyter, I was able to go back to the MacPorts Emacs version and also use emacs-jupyter normally. Any idea what happened?

**Comment by dzop** [comment]
In the case where you only have a GUI Emacs without a command line Emacs available you will have to do

```
(setenv "EMACS" "/Applications/Emacs.app/Contents/MacOS/Emacs -nw")
```

where `-nw` means to not open a GUI window.

> Any idea what happened?

The original error you were having was due to Emacs not being able to byte compile the `zmq.el` file of `emacs-zmq`, this byte compilation happens *after* compiling the C module. So on the second go around, the C module was already built.

**Comment by Salompas** [comment]
@dzop 
Thanks for your help!!!!! I appreciate your quick replies.

**Comment by eph** [comment]
I am also getting this error with emacs26.1 on Ubuntu 16.04.  It seems like zmq is building correctly: ```emacs-zmq.so``` exists and so does ```zmq.elc```.  Also, ```cl-lib``` is installed and ```module-file-suffix``` is set to ".so."  Any thoughts on what I might try?  This package looks extremely cool.

**Comment by dzop** [comment]
Can you show the exact error backtrace by doing `M-x toggle-debug-on-error` before the error happens. If both `emacs-zmq.so` and `zmq.elc` are both being built, then maybe `zmq` is not being found in your `load-path` when `(require 'zmq)` happens?

**Comment by eph** [comment]
Thanks for your reply!  The compile errors aren't providing a traceback for me (they happen when the files are being byte-compiled).  If I do ```(require 'jupyter)```, I get 

```
Debugger entered--Lisp error: (void-variable zmq-REQ)
  (list :hb zmq-REQ :shell zmq-DEALER :iopub zmq-SUB :stdin zmq-DEALER :control zmq-DEALER)
  (defconst jupyter-socket-types (list :hb zmq-REQ :shell zmq-DEALER :iopub zmq-SUB :stdin zmq-DEALER :control zmq-DEALER) ("/home/eherbst/.emacs.d/elpa/jupyter-0.7.2/jupyter-base.elc" . 3169))
  require(jupyter-base)
  eval-buffer(#<buffer  *load*> nil "/home/eherbst/.emacs.d/elpa/jupyter-0.7.2/jupyter.el" nil t)  ; Reading at buffer position 1236
  load-with-code-conversion("/home/eherbst/.emacs.d/elpa/jupyter-0.7.2/jupyter.el" "/home/eherbst/.emacs.d/elpa/jupyter-0.7.2/jupyter.el" nil t)
  require(jupyter)
  eval((require 'jupyter) nil)
  elisp--eval-last-sexp(nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  call-interactively(eval-last-sexp nil nil)
  command-execute(eval-last-sexp)
```

Doing ```(require 'zmq)``` doesn't give any error.

**Comment by dzop** [comment]
OK, can you go into the `emacs-zmq` directory and run `make test` from the command line to see if all the tests pass.

Also can you show what the version number of the `emacs-zmq` package you are using, the MELPA package version.

**Comment by eph** [comment]
Thanks again for helping me with this.
I am using version 20190210.1712 (commit 2246c5894dd2933d65312196e5064519444a92f5).

The file ```zmq-test.el``` is not in the directory.  The directory structure is 

```bash
├── emacs-zmq.la
├── emacs-zmq.so
├── libzmq
├── Makefile
├── src
├── version
├── zmq-autoloads.el
├── zmq.el
├── zmq.elc
└── zmq-pkg.el
```

ZMQ build log
[zmq-build-log.txt](https://github.com/dzop/emacs-jupyter/files/2882122/zmq-build-log.txt)

**Comment by dzop** [comment]
Oh whoops, I forgot that MELPA doesn't have the test file. Can you download it (https://raw.githubusercontent.com/dzop/emacs-zmq/master/zmq-tests.el) into the `emacs-zmq` directory and run the tests.

Also when you do `C-h v zmq- TAB` after `(require 'zmq)` do any `zmq` related variables pop up.

**Comment by eph** [comment]
Here is the output:

```
└──( make test
make test
emacs26 -nw -Q -batch -L . -l ert -l zmq-tests.el \
--eval "(ert-run-tests-batch-and-exit)"
Error in post-gc-hook (zmq--cleanup-globrefs): (void-function zmq--cleanup-globrefs)
Running 11 tests (2019-02-19 17:24:23-0500)
Test zmq-close-socket-once backtrace:
  (zmq-socket (zmq-current-context) zmq-PUB)
  (let ((sock (zmq-socket (zmq-current-context) zmq-PUB))) (zmq-close 
  (closure (t) nil (let ((sock (zmq-socket (zmq-current-context) zmq-P
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-close-socket-once :documentation 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-close-socket-once condition:
    (void-function zmq-socket)
   FAILED   1/11  zmq-close-socket-once
Test zmq-contexts backtrace:
  (zmq-context)
  (setq ctx (zmq-context))
  (let ((ert--infos (cons (cons "Info: " "Creating contexts") ert--inf
  (let (ctx) (let ((ert--infos (cons (cons "Info: " "Creating contexts
  (closure (t) nil (let (ctx) (let ((ert--infos (cons (cons "Info: " "
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-contexts :documentation nil :body
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-contexts condition:
    Info: Creating contexts
    (void-function zmq-context)
   FAILED   2/11  zmq-contexts
Test zmq-encryption backtrace:
  (zmq-has "curve")
  (if (zmq-has "curve") nil (ert-skip "ZMQ built without CURVE"))
  (closure (t) nil (if (zmq-has "curve") nil (ert-skip "ZMQ built with
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-encryption :documentation nil :bo
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-encryption condition:
    (void-function zmq-has)
   FAILED   3/11  zmq-encryption
Test zmq-globrefs backtrace:
  (zmq-poller)
  (let ((poller (zmq-poller))) (let ((sock (zmq-socket (zmq-current-co
  (let ((ert--infos (cons (cons "Info: " "`zmq-poller' socket referenc
  (let ((table (make-hash-table :size 1 :weakness 'value :test 'equal)
  (closure (t) nil (let ((table (make-hash-table :size 1 :weakness 'va
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-globrefs :documentation nil :body
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-globrefs condition:
    Info: `zmq-poller' socket references don't outlive poller
    (void-function zmq-poller)
   FAILED   4/11  zmq-globrefs
Test zmq-messages backtrace:
  (zmq-message)
  (let ((msg (zmq-message))) (let ((ert--infos (cons (cons "Info: " "E
  (let ((ert--infos (cons (cons "Info: " "Message initialization") ert
  (closure (t) nil (let ((ert--infos (cons (cons "Info: " "Message ini
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-messages :documentation nil :body
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-messages condition:
    Info: Message initialization
    (void-function zmq-message)
   FAILED   5/11  zmq-messages
Test zmq-polling backtrace:
  zmq-context()
  zmq-current-context()
  (let ((ctx (zmq-current-context))) (let ((ert--infos (cons (cons "In
  (closure (t) nil (let ((ctx (zmq-current-context))) (let ((ert--info
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-polling :documentation nil :body 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-polling condition:
    (void-function zmq-context)
   FAILED   6/11  zmq-polling
Test zmq-send-unicode backtrace:
  (zmq-message contents)
  (let* ((contents "[â, â†] = 1") (msg (zmq-message contents))) (let
  (let ((ert--infos (cons (cons "Info: " "Unicode messages") ert--info
  (closure (t) nil (let ((ert--infos (cons (cons "Info: " "Unicode mes
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-send-unicode :documentation nil :
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-send-unicode condition:
    Info: Unicode messages
    (void-function zmq-message)
   FAILED   7/11  zmq-send-unicode
Test zmq-sockets backtrace:
  (zmq-context)
  (let* ((ctx (zmq-context)) (endpoint "tcp://127.0.0.1:5345") (s (zmq
  (closure (t) nil (let* ((ctx (zmq-context)) (endpoint "tcp://127.0.0
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-sockets :documentation nil :body 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-sockets condition:
    (void-function zmq-context)
   FAILED   8/11  zmq-sockets
Test zmq-subprocess backtrace:
  signal(zmq-subprocess-error (void-function (zmq-context)))
  zmq--subprocess-filter(#<process zmq> "(error void-function (zmq-con
  sleep-for(0.01)
  (while (not filter-called) (sleep-for 0.01))
  (progn (while (not filter-called) (sleep-for 0.01)))
  (unwind-protect (progn (while (not filter-called) (sleep-for 0.01)))
  (let* ((-with-timeout-timer- (run-with-timer 0.4 nil (function (lamb
  (catch 'timeout (let* ((-with-timeout-timer- (run-with-timer 0.4 nil
  (let ((-with-timeout-value- (catch 'timeout (let* ((-with-timeout-ti
  (let* ((body '((if zmq-current-context (prin1 (cons 'test-result "co
  (let ((ert--infos (cons (cons "Info: " "Subprocess wraps function wi
  (closure (t) nil (let ((ert--infos (cons (cons "Info: " "Validating 
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-subprocess :documentation nil :bo
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-subprocess condition:
    Info: Subprocess wraps function with context
    (zmq-subprocess-error void-function
			  (zmq-context))
   FAILED   9/11  zmq-subprocess
Test zmq-utility backtrace:
  (zmq-version)
  (let ((version (zmq-version))) (let* ((fn-28 (function version<=)) (
  (let ((ert--infos (cons (cons "Info: " "`zmq-version'") ert--infos))
  (closure (t) nil (let ((ert--infos (cons (cons "Info: " "`zmq-versio
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name zmq-utility :documentation nil :body 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  eval((ert-run-tests-batch-and-exit))
  command-line-1(("-L" "." "-l" "ert" "-l" "zmq-tests.el" "--eval" "(e
  command-line()
  normal-top-level()
Test zmq-utility condition:
    Info: `zmq-version'
    (void-function zmq-version)
   FAILED  10/11  zmq-utility
   passed  11/11  zmq-wrong-object-type

Ran 11 tests, 1 results as expected, 10 unexpected (2019-02-19 17:24:38-0500)

10 unexpected results:
   FAILED  zmq-close-socket-once
   FAILED  zmq-contexts
   FAILED  zmq-encryption
   FAILED  zmq-globrefs
   FAILED  zmq-messages
   FAILED  zmq-polling
   FAILED  zmq-send-unicode
   FAILED  zmq-sockets
   FAILED  zmq-subprocess
   FAILED  zmq-utility
```

zmq-[TAB}
```
zmq
zmq-message-properties
zmq--subprocess-debug
zmq-module-file
zmq-backtrace
zmq-current-context
zmq--subprocess-parse-state
```

**Comment by dzop** [comment]
It looks like the `emacs-zmq.so` library is not being loaded. All of the `zmq-socket` and `zmq-context` functions and `zmq-REQ` variables are defined in the `emacs-zmq.so` library and the variables that you see when you do `C-h v zmq- TAB` are defined in `zmq.el`.

What is the value of `zmq-module-file`? And if you run `(zmq-load)` what happens?

**Comment by eph** [comment]
It looks like it's finding the file:
```
zmq-module-file is a variable defined in ‘zmq.el’.
Its value is
"/home/eherbst/.emacs.d/elpa/zmq-20190210.1712/emacs-zmq.so"

  This variable may be risky if used as a file-local variable.

Documentation:
The module file for ZMQ or nil if modules are not supported.
```

When I run ```(zmq-load)``` the status buffer displays ```(zmq--cleanup-globrefs)``` but this function does not exist in the global namespace (?, sorry not a lisp programmer), even after ```(require zmq)```.

**Comment by dzop** [comment]
OK, yeah it does look like it finds the file. Can you evaluate `(module-load zmq-module-file)` and see if you have more variables available whenever you do `C-h v zmq- TAB`.

**Comment by eph** [comment]
```(module-load zmq-module-file)``` returns t but I don't have any additional variables after executing the function.

**Comment by dzop** [comment]
Then it must have something to do with the Emacs that you are using since `module-load` is supposed to call into the `emacs-zmq.so` library and load all the definitions. Is there a newer Emacs that you can install?

**Comment by eph** [comment]
I'll try to figure that out --thanks!

**Comment by dzop** [comment]
Although Emacs 26 is the minimum supported version, you can install an Emacs 25 with module support and see if zmq works there. You can run the test like

```
make EMACS=emacs25 test 
```

and see if that works. This way you can see if it is more likely to be an Emacs issue.

**Comment by mvarela** [comment]
Hi, I seem to be having the same issue, using Emacs 26.2. 
Running `(module-load zmq-module-file)` returns `t`, but `zmq-REQ` remains undefined.

If I try to run it under HEAD, it just crashes while loading `zmq`.

**Comment by dzop** [comment]
@mvarela This could have to do with compiling Emacs itself. See https://github.com/dzop/emacs-zmq/issues/9. In that issue, a user was getting similar void function errors on Emacs 26.1.92, Ubuntu 16.04.

Can you try building Emacs with `CFLAGS=-O0` or at commit https://github.com/emacs-mirror/emacs/commit/9c0fa11 to see if you are able to get it working like the user in the issue did. If those don't work can you try running Emacs under `gdb` using some of the commands mentioned in https://github.com/dzop/emacs-zmq/issues/9 and show their output.

**Comment by mvarela** [comment]
I've just build 26.2 from source, using the `-O0` flag, and I still get the error about `zmq-REQ`'s value as a variable being void. 

Doing `(module-load zmq-module-file)` returns `t`, as before, so the module is loading. I'll keep looking into this, and try to make it work (I did manage to get emacs-jupyter working on mac, but I'm having some issues there with some Julia packages, so no joy, either...)

**Comment by dzop** [comment]
Can you try

```
emacs -nw -Q -batch --module-assertions -L . -l zmq --eval "(zmq-context)"
```

**Comment by mvarela** [comment]
I rebuilt the HEAD version of Emacs (and zmq) with clang, and that seems to have solved the problem, i can now use jupyter.

**Comment by dickmao** [comment]
I am wondering if users of `emacs-jupyter` are largely windows-based.
On ubuntu linux, I've compiled 25.3, 26.1, and 26.2 with modules support, and only 25.3 successfully pulled the zmq-REQ symbol.  I can't make heads or tails of it.

**Comment by dzop** [comment]
@dickmao It most likely has to do with compiling Emacs on Ubuntu. What compiler (and version of it) are you using and what version of Ubuntu are you on? Also see https://github.com/dzop/emacs-jupyter/issues/22#issuecomment-487785310.

There currently are no Windows users of `emacs-jupyter` since I have just recently been able to get Windows working in the `fix/windows-support` branch.

**Comment by mvarela** [comment]
@dickmao, I'm using Ubuntu 18.04, and was using the emacs-snapshot PPA. I managed to solve this by building Emacs with clang (building with gcc and optimizations disabled didn't help), but neither the emacs-snapshot nor the stock emacs25 version that ships with Ubuntu worked for me. Otherwise, building from source under Ubuntu was a PITA (pkg-config issues with gtk3, fetching all the deps, etc). 

So I don't know about windows users, but Ubuntu is a quite popular distro, and it seems problematic here...

**Comment by dzop** [comment]
I found [this](https://lists.gnu.org/archive/html/emacs-devel/2018-09/msg00069.html) thread on `emacs-devel` which shows that compiling on Ubuntu 16.04 using gcc 5.4.0 with optimizations causes issues with emacs module code. The issue in the `emacs-devel` thread is related to https://github.com/dzop/emacs-zmq/issues/9, both of them being a problem with calling the function `module_intern` in the `emacs-module.c` file of the Emacs sources.

**Comment by dickmao** [comment]
> I'm using Ubuntu 18.04, and was using the emacs-snapshot PPA

16.04, gcc 5.4.0, and emacs-snapshot PPA here.  Like you, after ppa didn't work, started compiling from github source to no avail.

> The issue in the emacs-devel thread is related to dzop/emacs-zmq#9, both of them being a problem with calling the function module_intern in the emacs-module.c file of the Emacs sources.

Yes, I experienced dzop/emacs-zmq#9 as well.  I've been hacking EIN for the last 6 months, and wanted to see if some messaging architecture could be shared.  Unfortunately, my personal requirements include, for now, compatibility with emacs25 and out-of-the-box operability.

**Comment by arthurcgusmao** [comment]
I use Ubuntu 16.04 and I run into the same issues mentioned before. Thanks @mvarela for mentioning the solution of building with clang (`$ ./configure --with-modules CC=clang` when building from source). I have detailed the steps I followed to get this package working [here](https://github.com/arthurcgusmao/acg-emacs/blob/master/others/linux/README.md). Thank you @dzop for the support and for this great package.
